crates/polars-utils/src/cache.rs

```diff
@@ -29,3 +38,197 @@ where
+        unsafe { self.raw_insert(self.hash(&key), key, value) }
+    }
+
+    pub fn get_or_insert_with<Q, F>(&mut self, key: &Q, f: F) -> &mut V
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ToOwned<Owned = K> + ?Sized,
+        F: FnOnce(&K) -> V,
+    {
+        unsafe {
+            let h = self.hash(key);
+            if let Some(slot_idx) = self.raw_get(self.hash(&key), key) {
+                let slot = self.slots.get_unchecked_mut(slot_idx);
+                return slot.value.assume_init_mut();
+            }
+
+            let key = key.to_owned();
+            let val = f(&key);
+            self.raw_insert(h, key, val)
+        }
+    }
+
+    unsafe fn raw_get<Q: Eq + ?Sized>(&self, h: HashResult, key: &Q) -> Option<usize>
+    where
+        K: Borrow<Q>,
+    {
+        unsafe {
+            // SAFETY: we assume h is a HashResult from self.hash with valid indices
+            // and we check slot.last_access != 0 before assuming the slot is initialized.
+            let slot = self.slots.get_unchecked(h.i1);
+            if slot.last_access.get() != 0
+                && slot.hash_tag == h.tag
+                && slot.key.assume_init_ref().borrow() == key
+            {
+                slot.last_access.set(self.new_access_ctr());
+                return Some(h.i1);
+            }
+
+            let slot = self.slots.get_unchecked(h.i2);
+            if slot.last_access.get() != 0
+                && slot.hash_tag == h.tag
+                && slot.key.assume_init_ref().borrow() == key
+            {
+                slot.last_access.set(self.new_access_ctr());
+                return Some(h.i2);
+            }
+        }
+
+        None
+    }
+
+    unsafe fn raw_insert(&mut self, h: HashResult, key: K, value: V) -> &mut V {
+        let last_access = self.new_access_ctr();
+        unsafe {
+            // SAFETY: i1 and i2 are valid indices and older_idx returns one of them.
+            let idx = self.older_idx(h.i1, h.i2);
+            let slot = self.slots.get_unchecked_mut(idx);
+
+            // Drop impl takes care of dropping old value, if occupied.
+            *slot = CacheSlot {
+                last_access: Cell::new(last_access),
+                hash_tag: h.tag,
+                key: MaybeUninit::new(key),
+                value: MaybeUninit::new(value),
+            };
+            slot.value.assume_init_mut()
+        }
+    }
+
+    /// Returns the older index based on access time, where unoccupied slots
+    /// are considered infinitely old.
+    unsafe fn older_idx(&mut self, i1: usize, i2: usize) -> usize {
+        let age1 = self.slots.get_unchecked(i1).last_access.get();
+        let age2 = self.slots.get_unchecked(i2).last_access.get();
+        match (age1, age2) {
+            (0, _) => i1,
+            (_, 0) => i2,
+            // This takes into account the wrap-around of our access_ctr.
+            _ if age1.wrapping_sub(age2) >= (1 << 31) => i1,
```

Comment:
I don't understand. How can we know `access_ctr` has wrapped around and accommodate for that?

Reply:
We don't know whether it has wrapped around, and it's not perfect. Consider if we had a 3-bit counter (and let's ignore my 'odd counter will never hit 0' trick for now). Then with wraparound it would look something like this:
![image](https://github.com/pola-rs/polars/assets/202547/0a0950de-c4d2-41e7-9845-fabb3566f042)

Now suppose we get two ages, `0` and `6`. Which is older?
![image](https://github.com/pola-rs/polars/assets/202547/62c850d9-2f2e-4c76-9ca5-059b132bead3)

The assumption I make is that generally speaking our counter doesn't increase that fast relative to how often values are accessed. So I will assume that the shortest possible distance between the two ages is 'the truth'. So in this case want that `6` is older than `0`. How do we get that? Well, if we look at the case where one of the numbers is 0, we see the following:

![image](https://github.com/pola-rs/polars/assets/202547/1dfa8302-1c1f-43fe-9490-c3b0cbf1a519)

In the red zones `0` is older than the other value, in the green zones the other value is older than `0`. We can generalize this to not just compare to `0` by shifting along the wrapping number line until `y` becomes zero: `x2 = x.wrapping_sub(y), y2 = y.wrapping_sub(y) = 0`. Combine it all and you get that `x` is older than `y` if `x.wrapping_sub(y) >= (1 << (counter_bits - 1))`.

Reply:
I see. Nice trick indeed. Can you add a link to this comment in the code? Future me will not remember this. ;)

Reply:
@ritchie46  I added two comments in the latest commit. Does that satisfy you?

Reply:
Yes. it does. :)